movement/kalman_filter.py

```diff
@@ -0,0 +1,215 @@
+"""Kalman filter implementation for motion tracking and smoothing."""
+
+import importlib
+import sys
+
+import h5py
+import numpy as np
+import xarray as xr
+from filterpy.kalman import KalmanFilter
+from scipy.spatial.distance import mahalanobis
+
+from movement import sample_data
+from movement.visualization import (
+    animate_motion_tracks,
+)
+
+# Move all imports to the top as per best practice
+sys.path.pop(0)  # Remove the first entry (which is your project folder)
+napari = importlib.import_module("napari")  # ✅ Napari stays integrated!
+
+
+def load_sleap_data_xarray(file_path):
+    """Load SLEAP data into an xarray Dataset with error handling.
+
+    Args:
+        file_path (str): Path to the SLEAP HDF5 file.
+
+    Returns:
+        xarray.Dataset:contains 'x' and 'y' coordinates for each frame.
+
+    """
+    try:
+        with h5py.File(file_path, "r") as hf:
+            poses = hf["tracks"][:]
+    except FileNotFoundError as err:
+        raise FileNotFoundError(
+            f"Dataset file {file_path} not found."
+        ) from err
+
+    # Squeeze and transpose to shape (frames, coordinates)
+    poses = np.squeeze(poses, axis=2).transpose(1, 0, 2)
+
+    dataset = xr.Dataset(
+        {"x": ("frame", poses[:, 0, 0]), "y": ("frame", poses[:, 0, 1])},
+        coords={"frame": np.arange(len(poses))},
+    )
+    return dataset
+
+
+def preprocess_xarray_data(dataset):
```

Comment:
Again, there is no need to define your own function for interpolating data.

We hava an built-in [interpolation function](https://movement.neuroinformatics.dev/api/movement.filtering.interpolate_over_time.html#movement.filtering.interpolate_over_time), which will work on properly formatted `movement` datasets.